src/main/java/org/jenkinsci/main/modules/sshd/PublicKeyAuthenticatorImpl.java

```diff
@@ -20,27 +28,74 @@
     private final PublicKeySignatureWriter signatureWriter = new PublicKeySignatureWriter();
 
     public boolean authenticate(String username, PublicKey key, ServerSession session) {
-        LOGGER.fine("Authentication attempted from "+username+" with "+key);
-        User u = User.get(username, false);
-        if (u==null) {
-            LOGGER.fine("No such user exists: "+username);
+        User user = this.retrieveOnlyKeyValidatedUser(username, key);
+
+        if (user == null) {
+            SecurityListener.fireFailedToAuthenticate(username);
             return false;
         }
 
-        UserPropertyImpl sshKey = u.getProperty(UserPropertyImpl.class);
-        if (sshKey==null) {
-            LOGGER.fine("No SSH key registered for user: "+username);
+        Authentication auth = this.verifyUserUsingSecurityRealm(user);
+        if (auth == null) {
+            SecurityListener.fireFailedToAuthenticate(username);
             return false;
```

Comment:
IIUC compared to the current behavior, from the perspective of the client this will silently reject the login attempt (akin to a bad/missing private key, etc.) rather than displaying an error making it clear that the problem is with the security realm rather than Jenkins per se.

Probably OK, but we may need to ensure that [documentation](https://jenkins.io/doc/book/managing/cli/#authentication) includes a troubleshooting guide explaining to turn on this `LOGGER` to diagnose puzzling authentication failures.

Reply:
Actually the previous code is still called `User user = this.retrieveOnlyKeyValidatedUser(username, key);`, so if you try to connect with a bad / missing private key, you will receive exactly the same message as currently.

Reply:
> if you try to connect with a bad / missing private key, you will receive exactly the same message as currently

Right. I am talking about the case that the user is not present in the external security realm, for which this change reduces self-diagnosability.

Reply:
PR on Jenkins.io for the related paragraphs: (#1202)[https://github.com/jenkins-infra/jenkins.io/pull/1202]